File: gameserver/src/Bot.ts
import { BotPos, EnginePower } from "./BotPhysics";

export default class Bot {
    x: number;
    y: number;
    angle: number;
    private _power: () => EnginePower;

    constructor(x = 0,y = 0, angle = 0){
        this.x = x;
        this.y = y;
        this.angle = angle;

        this._power = () => {return {
            left: 0,
            right: 0,
        }};
    }


    goTo(x: number, y: number): void {
        this._power = () => {
            let dy = y - this.y;
            let dx = x - this.x;
            let direction = Math.atan2(dy,dx);
            let offset = Math.atan2(Math.sin(direction-this.angle), Math.cos(direction-this.angle));
            let right = 1;
            let left = 1;
            if(offset > 0 ){
                left = offset / Math.PI
            }
            if(offset < 0) {
                right = -offset / Math.PI
            }
            let distance = Math.sqrt(dy*dy + dx * dx);
            let maxPower = Math.min(distance, 1);
            
            return {
                left: left * maxPower,
                right: right * maxPower,
            }
        }

    }

    turnToPoint(x: number, y: number): void {
        let dy = y - this.y;
        let dx = x - this.x;
        this.turnToAngle(Math.atan2(dy,dx));
    }

    turnToAngle(target: number): void {
        this._power = () => {
            let offset = Math.atan2(Math.sin(target-this.angle), Math.cos(target-this.angle));
            let right = offset / Math.PI;
            return {
                left: -right,
                right: right,
            }
        }
    }

    stop(): void {
        this.power = {
            left: 0,
            right: 0,
        }
    }

    get pos(): BotPos {
        return {
            x: this.x,
            y: this.y,
            angle: this.angle
        }
    }

    set pos(pos: BotPos){
        this.x = pos.x;
        this.y = pos.y;
        this.angle = pos.angle;
    }

    get power() : EnginePower {
        return this._power();
    }

    set power(power: EnginePower){
        this._power = () => power;
    }
}

File: putlebot-server/src/models/bot.rs
use crate::bots::bot_pos::BotPos;
use crate::bots::engine_power::EnginePower;
use crate::models::point::Point;
use crate::models::angle::Angle;
use std::f64::consts::PI;

#[derive(Debug, Clone, Copy)]
pub enum Action {
    GoTo(Point),
    TurnTo(Point),
    Power(EnginePower),
    Idle
}

#[derive(Debug, Clone, Copy)]
pub struct Bot{
    pos: BotPos,
    action: Action
}

impl Bot{
    pub fn new(pos: BotPos) -> Bot {
        Bot {
            pos,
            action: Action::Idle
        }
    }

    pub fn set_pos(self, pos: BotPos) -> Bot {
        Bot {
            pos,
            action: self.action
        }
    }

    pub fn power(self) -> EnginePower {
        match self.action {
            Action::GoTo(p) => self.go_to_calc(p),
            Action::TurnTo(p) => self.turn_to_calc(p),
            Action::Power(p) => p,
            Action::Idle => EnginePower::IDLE
        }
    }

    pub fn set_power(self, power: EnginePower) -> Bot {
        Bot{
            pos: self.pos,
            action: Action::Power(power),
        }
    }

    pub fn go_to(self, target: Point) -> Bot {
        Bot{
            pos: self.pos,
            action: Action::GoTo(target)
        }
    }

    pub fn turn_to(self, target: Point) -> Bot {
        Bot{
            pos: self.pos,
            action: Action::TurnTo(target)
        }
    }
    
    pub fn stop(self) -> Bot {
        Bot{
            pos: self.pos,
            action: Action::Idle
        }
    }

    fn go_to_calc(self, target: Point) -> EnginePower {
        let delta = target.sub(self.pos.point());
        let offset = Angle::from_point(delta).sub(self.pos.angle()).right() * 0.5;
        let mut right = 0.5;
        let mut left = 0.5;
        if offset > 0.0 {
            left = -offset / PI
        }
        if offset < 0.0  {
            right = offset / PI
        }
        let max_power = delta.distance().min(1.0);
        
        EnginePower::new(left * max_power, right * max_power)
    }

    fn turn_to_calc(self, target: Point) -> EnginePower {
        let target = Angle::from_point(target.sub(self.pos.point()));
        let right = target.sub(self.pos.angle()).right() * 0.5;
        EnginePower::new(-right, right)
    }
}